cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/SourceControlManagementHttpHandler.java

```diff
@@ -163,6 +171,38 @@ public void pushApp(FullHttpRequest request, HttpResponder responder,
     }
   }
 
+  /**
+   * Pull the requested application from linked repository and deploy in current namespace.
+   */
+  @POST
+  @Path("/apps/{app-id}/pull")
+  public void pullApp(FullHttpRequest request, HttpResponder responder,
+                      @PathParam("namespace-id") String namespaceId,
+                      @PathParam("app-id") final String appId) throws Exception {
+    checkSourceControlFeatureFlag();
+    ApplicationReference appRef = validateAppReference(namespaceId, appId);
+
+    try {
+      ApplicationRecord appRecord = sourceControlService.pullAndDeploy(appRef);
+      responder.sendJson(HttpResponseStatus.OK, GSON.toJson(appRecord));
+    } catch (NoChangesToPullException | InvalidArtifactException e) {
```

Comment:
Should we return 400 or 200 if we found there's no change for the application to pull?

Reply:
400 usually means there is something wrong with the request itself, such that it will always be wrong and always result in a 400 on retries.

I want to say this is similar to when there is a PUT API and you put the same value. In that case it's still a 200 even if it's a no-op.

When does NoChangesToPull get thrown? If the current version's filehash is the same as the one in the remote git repository?

Reply:
Sounds great, now returning 200 to make it idempotent